lib/charms/tls_certificates_interface/v3/tls_certificates.py

```diff
@@ -687,6 +691,68 @@ def _get_certificate_expiry_time(certificate: str) -> Optional[datetime]:
         return None
 
 
+def _get_certificate_validity_start_time(certificate: str) -> Optional[datetime]:
+    """Extract expiry time from a certificate string.
+
+    Args:
+        certificate (str): x509 certificate as a string
+
+    Returns:
+        Optional[datetime]: Time when this certificate is valid from or None
+    """
+    try:
+        certificate_object = x509.load_pem_x509_certificate(data=certificate.encode())
+        return certificate_object.not_valid_before_utc
+    except ValueError:
+        logger.warning("Could not load certificate.")
+        return None
+
+
+def _calculate_expiry_notification_time(
+        validity_time: datetime,
+        expiry_time: datetime,
+        provider_recommended_notification_time: Optional[int],
+        requirer_recommended_notification_time: Optional[int],
+    ) -> datetime:
+        """Calculate a reasonable time to notify the user about the certificate expiry.
+
+        It takes into account the time recommended by the provider and by the requirer.
+        Time recommended by the provider is preferred,
+        then time recommended by the requirer,
+        then dynmaicaly calculated time.
+
+        Args:
+            validity_time: Certificate validity time
+            expiry_time: Certificate expiry time
+            provider_recommended_notification_time:
```

Comment:
Please specify the unit (seconds, minutes, hours?), same for the `requirer_recommended_notification_time`

Reply:
It is specified in the docstring, next line was `Time in hours prior to expiry to notify the user.`

Reply:
Oops, my bad